runtime/drivers/file/watcher.go

```diff
@@ -0,0 +1,224 @@
+package file
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"path/filepath"
+	"sync"
+	"time"
+
+	"github.com/fsnotify/fsnotify"
+	runtimev1 "github.com/rilldata/rill/proto/gen/rill/runtime/v1"
+	"github.com/rilldata/rill/runtime/drivers"
+)
+
+const batchInterval = 250 * time.Millisecond
+
+// watcher implements a recursive, batching file watcher on top of fsnotify.
+type watcher struct {
+	root        string
+	watcher     *fsnotify.Watcher
+	done        chan struct{}
+	err         error
+	mu          sync.Mutex
+	subscribers map[string]drivers.WatchCallback
+	buffer      []drivers.WatchEvent
+}
+
+func newWatcher(root string) (*watcher, error) {
+	fsw, err := fsnotify.NewWatcher()
+	if err != nil {
+		return nil, err
+	}
+
+	w := &watcher{
+		root:        root,
+		watcher:     fsw,
+		done:        make(chan struct{}),
+		subscribers: make(map[string]drivers.WatchCallback),
+	}
+
+	err = w.addDir(root, false)
+	if err != nil {
+		w.watcher.Close()
+		return nil, err
+	}
+
+	go w.run()
+
+	return w, nil
+}
+
+func (w *watcher) close() {
+	w.closeWithErr(nil)
+}
+
+func (w *watcher) closeWithErr(err error) {
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	select {
+	case <-w.done:
+		// Already closed
+		return
+	default:
+	}
+
+	closeErr := w.watcher.Close()
+	if err != nil {
+		w.err = err
+	} else if closeErr != nil {
+		w.err = closeErr
+	} else {
+		w.err = fmt.Errorf("file watcher closed")
+	}
+
+	close(w.done)
+}
+
+func (w *watcher) subscribe(ctx context.Context, fn drivers.WatchCallback) error {
+	w.mu.Lock()
+	if w.err != nil {
+		w.mu.Unlock()
+		return w.err
+	}
+	id := fmt.Sprintf("%v", fn)
+	w.subscribers[id] = fn
+	w.mu.Unlock()
+
+	defer func() {
+		w.mu.Lock()
+		delete(w.subscribers, id)
+		w.mu.Unlock()
+	}()
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-w.done:
+		return w.err
+	}
+}
+
+// flush emits buffered events to all subscribers.
+// Note it is called in the event loop in runInner, so new events will not be appended to w.buffer while a flush is running.
+// Calls to flush block until all subscribers have processed the events. This is an acceptable trade-off for now, but we may want to revisit it in the future.
+func (w *watcher) flush() {
+	if len(w.buffer) == 0 {
+		return
+	}
+
+	w.mu.Lock()
+	defer w.mu.Unlock()
+
+	for _, fn := range w.subscribers {
+		fn(w.buffer)
+	}
+
+	w.buffer = nil
+}
+
+func (w *watcher) run() {
+	err := w.runInner()
+	w.closeWithErr(err)
+}
+
+func (w *watcher) runInner() error {
+	timer := time.NewTimer(batchInterval)
+	timerActive := true
+	for {
+		select {
+		case <-timer.C:
+			timerActive = false
+			w.flush()
+		case err, ok := <-w.watcher.Errors:
+			if !ok {
```

Comment:
Decided not to flush in error scenarios since something with the underlying file access might have gone wrong, so we don't want to trigger actions against the buffered paths